File: Leetcode 0200 Number of Islands [medium] [array]/Number of Islands [dfs].java
class Solution {
    int[][] dirs = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}};
    
    public int numIslands(char[][] grid) {
        if (grid == null || grid.length == 0 || grid[0].length == 0) {
            return 0;
        }
        
        int rows = grid.length;
        int cols = grid[0].length;
        int count = 0;
        
        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
                if (grid[i][j] == '1') {
                    count++;
                    dfs(grid, i, j, rows, cols);
                }
            }
        }
        
        return count;
    }
    
    private void dfs(char[][] grid, int x, int y, int rows, int cols) {
        if (x < 0 || x >= rows || y < 0 || y >= cols || grid[x][y] == '0') {
            return;
        }
        
        grid[x][y] = '0';
        
        for (int[] dir : dirs) {
            int neighborX = x + dir[0];
            int neighborY = y + dir[1];
            dfs(grid, neighborX, neighborY, rows, cols);
        }
    }
}

File: Leetcode 0200 Number of Islands [medium] [array]/Number of Islands [dfs].js
/**
 * @param {character[][]} grid
 * @return {number}
 */
let dirs = [[-1, 0], [1, 0], [0, -1], [0, 1]];

var numIslands = function(grid) {
    if (grid === null || grid.length === 0 || grid[0].length === 0) {
        return 0;
    }
    
    let count = 0;
    let rows = grid.length;
    let cols = grid[0].length;
    for (let i = 0; i < rows; i++) {
        for (let j = 0; j < cols; j++) {
            if (grid[i][j] === '1') {
                count++;
                dfs(grid, i, j, rows, cols);
            }
        }
    }
    
    return count;
};

let dfs = (grid, x, y, rows, cols) => {
    if (x < 0 || x >= rows || y < 0 || y >= cols || grid[x][y] === '0') {
        return;
    }
    console.log(dirs);
    
    grid[x][y] = '0';
    for (let dir of dirs) {
        let neighborX = x + dir[0];
        let neighborY = y + dir[1];
        dfs(grid, neighborX, neighborY, rows, cols);
    }
}
    